### ISSUE TITLE
Rule proposal: `prefer-reflect-apply`

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/eslint-plugin-unicorn#142]
[<img alt="Issuehunt badges" src="https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/142/badge.svg" />](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/142)

ESLint deprecated the [`prefer-reflect`](https://eslint.org/docs/rules/prefer-reflect) rule, for mostly good reason, but there was one valid thing in it, and that's using `Reflect.apply()` over `Function.prototype.apply()`. When you accept arbitrary methods, it's not safe to assume `.apply()` exists or is not overridden.

Relevant discussion: https://github.com/eslint/eslint/issues/7226#issuecomment-257108713

## Fail

```js
function foo() {}

foo.apply(null, [true]);

Function.prototype.apply.call(foo, null, [true]);
```

## Pass

```js
function foo() {}

Reflect.apply(foo, null, [true]);
```

> There is a $80.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/issues/142).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/eslint-plugin-unicorn/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

**Comment by IssuehuntBot** [comment]
@issuehuntfest has funded $80.00 to this issue. [See it on IssueHunt](https://issuehunt.io/repos/55832243/issues/142)

**Comment by sindresorhus** [comment]
> foo.apply(null, [true]); would probably cause a lot of false positives since it's hard to know what foo is.

I'm not sure it will. I don't think it's likely a custom method is named `apply`, has `null` as the first argument, and an array as the second argument. I think we're fine here. It's the best we can do with [linting in JS](https://twitter.com/sindresorhus/status/1089602422236864512).

> Reflect.apply is not available at all in IE... not sure how much we care 🤔

That's not a concern. If people need to support IE, they can just disable the rule.

**Comment by futpib** [comment]
> Oh ok, I guess looking for `null` + array literal combination should be enough for most cases.

I don't think so. I bet the most common use-case for `apply` is an old-school variant of `super.method(...args)`, that is `method.apply(this, arguments)`;

**Comment by futpib** [comment]
Can we make this rule fix all `.apply` calls with 2 arguments instead of having it look for `null` and an array literal (what seems to be the current plan)?

**Comment by alexzherdev** [comment]
This seems too broad to me, but I'll defer to you and @sindresorhus.

**Comment by sindresorhus** [comment]
I also think that's too broad, but you made me realize we left out some common patterns. I think we should also handle `.apply` when the first argument is `this` and there's a second argument (of anything). @futpib Can you think of any other use-cases that would not be handled by that?

**Comment by futpib** [comment]
> Can you think of any other use-cases that would not be handled by that?

Not really, the next use-case on my list is the most generic `.apply(anything, anythingElse)`. But why do you think handling it is too broad? Do you know of any usage of `apply` method name with a meaning different from that of `Function.prototype.apply` or `Reflect.apply`?

**Comment by sindresorhus** [comment]
> But why do you think handling it is too broad? Do you know of any usage of `apply` method name with a meaning different from that of `Function.prototype.apply` or `Reflect.apply`?

I don't. I just know from experience that it's beneficial to be careful about these things.

@futpib @alexzherdev How about we report on any usage of `.apply()` with arguments, but only auto-fix the safe cases (the already mentioned cases above)?

**Comment by alexzherdev** [comment]
Coming from eslint-plugin-react I have an instinct to avoid false positives completely. That said, if this repo is more lenient, I'm totally cool with that, and this suggestion seems like a reasonable middle ground.

**Comment by futpib** [comment]
Let's report `.apply` with 2 argument where 1st argument is either `null` or `this`.

> Do you know of any usage of apply method name with a meaning different from that of Function.prototype.apply or Reflect.apply?

I found one myself: [`async.apply`](https://caolan.github.io/async/docs.html#apply).

I originally leaned towards reporting any `.apply` call with 2 arguments. But considering the `async.apply` example and that this rule's report is more of a suggestion to upgrade rather than an indication of a possible error, I changed my mind.

I wish we could grep the JS world to answer these "safety" questions more definitely.
